bdns/servers.go

```diff
@@ -0,0 +1,159 @@
+package bdns
+
+import (
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// serverProvider represents a type which can provide a list of addresses for
+// the bdns to use as DNS resolvers. Different implementations may provide
+// different strategies for providing addresses, and may provide different kinds
+// of addresses (e.g. host:port combos vs IP addresses).
+type ServerProvider interface {
+	Addrs() ([]string, error)
+	Stop()
+}
+
+// staticProvider stores a list of host:port combos, and provides that whole
+// list in randomized order when asked for addresses. This replicates the old
+// behavior of the bdns.impl's servers field.
+type staticProvider struct {
+	servers []string
+}
+
+var _ ServerProvider = &staticProvider{}
+
+func NewStaticProvider(servers []string) *staticProvider {
+	return &staticProvider{servers: servers}
+}
+
+func (sp *staticProvider) Addrs() ([]string, error) {
+	if sp.servers == nil || len(sp.servers) == 0 {
+		return nil, fmt.Errorf("no servers configured")
+	}
+	r := make([]string, len(sp.servers))
+	perm := rand.Perm(len(sp.servers))
+	for i, v := range perm {
+		r[i] = sp.servers[v]
+	}
+	return r, nil
+}
+
+func (sp *staticProvider) Stop() {}
+
+// dynamicProvider uses DNS to look up the set of IP addresses which correspond
+// to its single host. It returns this list in random order when asked for
+// addresses, and refreshes it regularly using a goroutine started by its
+// constructor.
+type dynamicProvider struct {
+	host string
+	// A map of IP addresses (Target fields in SRV records) to ports (Port fields
+	// in SRV records) associated with those addresses.
+	addrs  map[string][]uint16
+	mu     sync.RWMutex
+	cancel chan interface{}
+}
+
+var _ ServerProvider = &dynamicProvider{}
+
+// StartDynamicProvider constructs a new dynamicProvider and starts its
+// auto-update goroutine. The auto-update process queries DNS for SRV records
+// at refresh intervals and uses the resulting IP/port combos to populate the
+// list returned by Addrs. The update process ignores the Priority and Weight
+// attributes of the SRV records.
+func StartDynamicProvider(server string, refresh time.Duration) (*dynamicProvider, error) {
+	if server == "" {
+		return nil, fmt.Errorf("no DNS host provided")
```

Comment:
Very tiny DNS terminology nit: this is a domain name, not a host.

Speaking of which, we should specify in the doccomment that the `server` param should contain e.g. `example.com`, which will result in a query to `_dns._udp.example.com`.

Reply:
Good call. I've changed this from "host" to "name" or "domain name" everywhere (I think), and fleshed out the doc a bit more.